pixl_core/tests/uploader/test_xnat.py

```diff
@@ -0,0 +1,193 @@
+
+    Once the server has started, you can log in by visiting http://localhost:8080
+    with the username and password set in the `XNAT_USER_NAME` and `XNAT_PASSWORD`
+    environment variables.
+    """
+    config = xnat4tests.Config(
+        xnat_port=os.environ["XNAT_PORT"],
+        docker_host=os.environ["XNAT_HOST"],
+        build_args={
+            "xnat_version": "1.8.10.1",
+        },
+    )
+    xnat4tests.start_xnat(config)
+
+    # Create the project as well as a non-admin user to perform the upload
+    with xnat.connect(
+        server=config.xnat_uri,
+        user="admin",
+        password="admin",  # noqa: S106
+    ) as session:
+        session.post(
+            path="/xapi/users/",
+            json=dict(  # noqa: C408
+                admin=False,
+                username=os.environ["XNAT_USER_NAME"],
+                password=os.environ["XNAT_PASSWORD"],
+                firstName="pixl",
+                lastName="uploader",
+                email="pixl-uploader@pixl",
+                verified=True,
+                enabled=True,
+            ),
+            headers={"Content-Type": "application/json"},
+            accepted_status=[201, 409],
+        )
+
+        # XNAT requires project metadata to be uploaded as XML
+        with (TEST_DIR / "data" / "xnat_project.xml").open() as file:
+            project_xml = file.read()
+        session.skip_response_content_check = (
+            True  # so XNATPy doesn't raise an error if the project exists
+        )
+        session.post(
+            path="/data/projects",
+            data=project_xml,
+            headers={"Content-Type": "application/xml"},
+            accepted_status=[200, 409],
+        )
+        session.put(
+            path=f"/data/projects/{xnat_study_tags.project_slug}/users/Owners/pixl",
+            accepted_status=[200],
+        )
+
+    yield config.xnat_uri
+
+    # If a test instance already exists, xnat4tests will reuse it for subsequent tests.
+    # This can save a lot of time when testing locally as it takes several minutes to start
+    # the XNAT server.
+    if os.environ.get("XNAT_DESTROY_INSTANCE", "False").lower() == "true":
+        xnat4tests.stop_xnat(config)
+        return
+
+    # If we're keeping the instance, we need to remove the data before next test run.
+    # We do not delete the project itself as Project IDs cannot be reused.
+    with xnat.connect(
+        server=config.xnat_uri,
+        user="admin",
+        password="admin",  # noqa: S106
+    ) as session:
+        project = session.projects[xnat_study_tags.project_slug]
+        for subject in project.subjects.values():
+            session.delete(
+                path=f"/data/projects/{project.id}/subjects/{subject.label}",
+                query={"removeFiles": "True"},
+            )
+
+
+@pytest.mark.usefixtures("xnat_server")
+def test_upload_to_xnat(zip_dicoms, xnat_uploader, xnat_study_tags) -> None:
```

Comment:
Hi Paul, 

Does this need to pass for me (see logs below)?
```
tests/uploader/test_xnat.py::test_upload_to_xnat FAILED                                                                                                                                                     [100%]

==================================================================================================== FAILURES =====================================================================================================
_______________________________________________________________________________________________ test_upload_to_xnat _______________________________________________________________________________________________

zip_dicoms = <_io.BytesIO object at 0x7c8f6d294450>, xnat_uploader = <test_xnat.MockXNATUploader object at 0x7c8f6d44cd10>
xnat_study_tags = StudyTags(pseudo_anon_image_id='1.3.6.1.4.1.14519.5.2.1.99.1071.12985477682660597455732044031486', project_slug='some-project-slug', patient_id='987654321')

    @pytest.mark.usefixtures("xnat_server")
    def test_upload_to_xnat(zip_dicoms, xnat_uploader, xnat_study_tags) -> None:
        """Tests that DICOM image can be uploaded to the correct location"""
>       xnat_uploader.upload_to_xnat(
            zip_content=zip_dicoms,
            study_tags=xnat_study_tags,
        )

tests/uploader/test_xnat.py:168: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/core/uploader/_xnat.py:90: in upload_to_xnat
    session.services.import_(
../../../../../mambaforge/envs/pixlVE/lib/python3.11/site-packages/xnat/services.py:246: in import_
    response = self.xnat_session.upload_stream(uri=uri, stream=data, query=query, content_type=content_type, method='post')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <xnat.session.XNATSession object at 0x7c8f6d7846d0>
uri = 'http://localhost:8080/data/services/import?overwrite=none&dest=/archive&project=some-project-slug&subject=987654321&session=1.3.6.1.4.1.14519.5.2.1.99.1071.12985477682660597455732044031486&import-handler=DICOM-zip'
stream = <_io.BytesIO object at 0x7c8f6d294450>, retries = 1, query = {'dest': '/archive', 'import-handler': 'DICOM-zip', 'overwrite': 'none', 'project': 'some-project-slug', ...}
content_type = 'application/zip', method = 'post', overwrite = False, timeout = None

    def upload_stream(self,
                      uri: str,
                      stream: Union[IO, io.BufferedIOBase, io.TextIOBase],
                      retries: int = 1,
                      query: Optional[Dict[str, str]] = None,
                      content_type: Optional[str] = None,
                      method: str = 'put',
                      overwrite: bool = False,
                      timeout: TimeoutType = None):
        """
        Upload path from a stream to XNAT
    
        :param uri: uri to upload to
        :param stream: the file handle, path to a file or a string of path
                      (which should not be the path to an existing file!)
        :param retries: amount of times xnatpy should retry in case of
                        failure
        :param query: extra query string content
        :param content_type: the content type of the file, if not given it will
                             default to ``application/octet-stream``
        :param method: either ``put`` (default) or ``post``
        :param overwrite: indicate if previous data should be overwritten
        :param timeout: timeout in seconds, float or (connection timeout, read timeout)
        """
        self._check_connection()
    
        if overwrite:
            if query is None:
                query = {}
            query['overwrite'] = 'true'
    
        uri = self._format_uri(uri, query=query)
        self.logger.info('UPLOAD URI {}'.format(uri))
        attempt = 0
        response = None
    
        # Set the content type header
        if content_type is None:
            headers = {'Content-Type': 'application/octet-stream'}
        else:
            headers = {'Content-Type': content_type}
    
        while attempt < retries:
            stream.seek(0)
            attempt += 1
    
            if method == 'put':
                response = self.interface.put(uri, data=stream, headers=headers, timeout=timeout)
            elif method == 'post':
                response = self.interface.post(uri, data=stream, headers=headers, timeout=timeout)
            else:
                raise ValueError('Invalid upload method "{}" should be either put or post.'.format(method))
    
            try:
                self._check_response(response)
                return response
            except exceptions.XNATResponseError:
                pass
    
        # We didn't return correctly, so we have an error
>       raise exceptions.XNATUploadError(f'Upload failed after {retries} attempts! Status code'
                                         f' {response.status_code}, response text {response.text}')
E       xnat.exceptions.XNATUploadError: Upload failed after 1 attempts! Status code 400, response text <html>
E       <head>
E          <title>Status page</title>
E       </head>
E       <body>
E       <h3>unable to archive for the Prearchive session</h3><p>You can get technical details <a href="http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.1">here</a>.<br>
E       Please continue your visit at our <a href="/">home page</a>.
E       </p>
E       </body>
E       </html>

../../../../../mambaforge/envs/pixlVE/lib/python3.11/site-packages/xnat/session.py:932: XNATUploadError
---------------------------------------------------------------------------------------------- Captured stdout setup ----------------------------------------------------------------------------------------------
                                                                                                                                             
```

Reply:
not sure what the issue was, but @mxochicale and I destroyed his running instance and reran the tests and they seem to pass (and pass when re-running the tests too). Might be an issue with the instance being created before I fixed the tests

Reply:
Thanks Paul, all good with `pixl_core/tests/uploader/test_xnat.p`, some commands that were used: 
```
rm -rf .xnat4tests
docker system prune
export XNAT_DESTROY_INSTANCE=false
```

Reply:
I wouldn't change any environmental variables but maybe type
```
docker system prune -a
```